### ISSUE TITLE
finger support

### ISSUE THREAD

**Issue by lrvick** [hashbang/hashbang#9]
We should find a way (Even if it is faked under the hood) to allow users to create .plan, .project, and .pgpkey files and have "finger someuser" actually work as expected.

This is a core part of the multi-user unix experience, and offers a fun/classic way of showing how people did "Personal profiles" and "microblogging" before Facebook/Twitter.

**Comment by lrvick** [comment]
Per https://tools.ietf.org/html/rfc1288 it looks like implementing our own finger server should be stupid simple, and can map to whatever user backend we want.

**Comment by RyanSquared** [comment]
:+1: once we get the new setup this could easily be done. Why create a new issue though?

**Comment by daurnimator** [comment]
> Why create a new issue though?

So that someone can come along and say "okay I'll implement a finger server"

**Comment by RyanSquared** [comment]
So, I could in theory implement the finger server, but I'd need some details first.
1. Is it possible to read a list of users from the database, without giving the server itself a higher acccess?
2. What files should I check for in the finger server?
   - This also includes how should I check them? Just check /home/<user>/<file> ?
   - What happens if I don't have permissions to read the file?
3. Is there an actually readable specification?
4. Should I limit it to just the local server or should I attempt to target the user's "home" server - the one specified in registration?

**Comment by KellerFuchs** [comment]
@ChickenNuggers What's the design you have in mind?

Regarding the [userdb](https://github.com/hashbang/userdb), there is a design spec and a DB schema, but no API spec so far.

**Comment by RyanSquared** [comment]
I can in theory design it with any restrictions put in place but I need to know those restrictions before I have a design.

**Comment by daurnimator** [comment]
> Regarding the userdb, there is a design spec and a DB schema, but no API spec so far.

What sort of API are you talking about here?
The DB schema implicitly gives us a sql based API.
If we use postgrest then we have 99% of a HTTP API done for us already.

**Comment by RyanSquared** [comment]
> What sort of API are you talking about here?

If I recall correctly, @lrvick wanted to design an API where all applications or daemons use the API rather than interacting with the database manually. I was personally against the idea as I feel an API rather than a library means more overhead.

**Comment by daurnimator** [comment]
I'm still not sure what sort of API you're talking about. a http api? a python api? some protocol over a unix socket? a dbus api?

**Comment by KellerFuchs** [comment]
> an API rather than a library means more overhead

@ChickenNuggers Sorry, but I have to call bullshit:
1. Which difference do you make between “an API” and the API implemented by your hypothetical lib?
2. There is exactly as much network overhead involved (you need to connect to the DB in any case)
3. Even if there was some overhead, how much is there and is it relevant?

@daurnimator The proposal, as far as I recall, was to have:
- an HTTP API to provide anonymous [create/read operations and statistics](https://github.com/hashbang/userdb/blob/master/DESIGN.md#hashbangsh);
- a local `hashbangctld` daemon, running on shell servers, that allows users to write to their own row, in the `data` column;  
  I wrote [a spec](https://gist.github.com/KellerFuchs/3ccdebe1fa466f55ffb5) for it, a while back, but it hasn't happened yet, in large part because the `#!ctl` code was not exactly “nice” to work with.

Note that this doesn't specify how `hashbangctld` writes to the DB (it can't write to the local one, the shell-servers have read-only replicas).

I would rather not expose the master pgsql to the Internet (IIRC, it only needs to _connect_ to the other instances, for replication), and PostgREST can likely serve as the write interface there.
However, it seems pretty hard to add extra validation there (checking if the SSH keys are valid, if the JSON data object adheres to some schema, ...), so I'm not sure it should be the public API.

I guess @lrvick should pitch in, in case I forgot things we discussed.

**Comment by RyanSquared** [comment]
> Which difference do you make between “an API” and the API implemented by your hypothetical lib?

Well, I'd need to implement a library to connect to the HTTP API which connects to the database instead of just connecting directly to the database.

> There is exactly as much network overhead involved (you need to connect to the DB in any case)

I need to connect to the database but using the HTTP API as a middle man means:
1. I can't guarantee that the HTTP API is providing reliable information, and
2. That's more bandwidth used and another interface that needs to be made
   - This assumes this is the hashbangctld interface
3. It's not much of a hit on the processing depending on the mechanism used; using HTTP/2.0 could require more processing power while HTTP/1.1 could require more bandwidth used

Anyways, the above might become irrelevant considering the IRC discussion last night has brought about some new information:
- A container will run on a server that can interface with the PGSQL database, with write access to the `finger` field of users.
- User's information in the files listed below will be sent to the above container when modified
  - `.gpgkey`
  - `.sshkey`
  - `.plan`
  - `.profile`
  - Files listed in a `.track` file
  - Any other standard `finger` files
- The PGSQL database synchronizes the information with all shell servers
  - This process should be automatic for all changes, not just changes relating to `fingerd`
- A user attempts to contact a `fingerd` on `hashbang.sh` or any of the shell servers
- The daemon contacts a readonly copy of the database on the local filesystem
  - It's not specified how this would work when contacting `hashbang.sh`
- The filesystem returns the user's information to the daemon
- The daeomon relays the information back to the user requesting the information

**Comment by KellerFuchs** [comment]
The whole point of not having applications talk directly to the DB is to have a shared abstraction layer, so that we can validate the data and migrate to new schema or backends in a single place, rather than having to care about all software that ever used the userdb.

Could you upload a log of that discussion?  I seem to have missed it.
In any case, extra feedback:

> The PGSQL database synchronizes the information with all shell servers.

Just to be clear: it only means that there is a read-only copy of the userdb available on the shell servers.  See [the specification](https://github.com/hashbang/userdb/blob/master/DESIGN.md#replication).
Unless you plan to run stuff on the shell servers, though, I don't see how that's relevant.

> A user attempts to contact a fingerd on hashbang.sh or any of the shell servers

I would be strongly in favour of something that does not have to be run on the shell servers.

> User's information in the files listed below will be sent to the above container

Any specific reason not to store the data directly in userdb?
Each user has an extensible `data` JSON object, so you could store things in `data.finger.plan`, `data.finger.profile` and what's not.

**Comment by daurnimator** [comment]
> The whole point of not having applications talk directly to the DB is to have a shared abstraction layer, so that we can validate the data and migrate to new schema or backends in a single place, rather than having to care about all software that ever used the userdb.

I'd like to do this _in_ the database.
- validation is _best_ to do in the database, I would not do it anywhere else
- easy to version your database schemas. 
- can do backwards compatibility with views

**Comment by KellerFuchs** [comment]
@daurnimator Yes.  If you read what I wrote, I was referring to what we _cannot_ do in the DB.
Also, if you recall, the design document says we should enforce as much as possible in the schema (and we do).
- Not all data constraints can sanely be expressed as DDL constraints.  
  In our case, they are likely easy-ish to express with an extension for JSON validation, but it doesn't exist yet.
- Views cannot let you achieve backwards compat when the DB engine is the thing being replaced.

**Comment by RyanSquared** [comment]
> I would be strongly in favour of something that does not have to be run on the shell servers.

That would remove the ability to `finger <user>` or `finger <user>@localhost` though.

> Any specific reason not to store the data directly in userdb?
> Each user has an extensible data JSON object, so you could store things in data.finger.plan, data.finger.profile and what's not.

The container stores it in the userdb; this way there's only one connection that's given write access to the userdb **and** it's not on the shell servers.

**Comment by KellerFuchs** [comment]
> That would remove the ability to finger <user> or finger <user>@localhost though.

No.  Local finger looks for the files directly in the homedir.

> The container stores it in the userdb; this way there's only one connection that's given write access to the userdb and it's not on the shell servers.

E&&@(_#YUO(PI(PRWU(_!@*#!!#$I)(#)%#@

Please read the fucking design doc at some point, thanks.

**Comment by daurnimator** [comment]
> No. Local finger looks for the files directly in the homedir.

FWIW I _don't_ want the finger data in my homedir. only in the userdb.

e.g. ssh keys: I keep them updated in the userdb. Why would I want to keep them updated in my homedir _too_?

**Comment by KellerFuchs** [comment]
> > No. Local finger looks for the files directly in the homedir.
> 
> FWIW I don't want the finger data in my homedir. only in the userdb.

Same: having multiple copy of the data means they will not be consistent.
Hence my remark about being sad we can't easily set a default domain for finger.

**Comment by daurnimator** [comment]
> we can't easily set a default domain for finger

On the shell servers we can easily provide a custom `finger` binary.

**Comment by KellerFuchs** [comment]
@daurnimator Fair point.  More annoying than it needs to be, but that's likely the best option.

**Comment by RyanSquared** [comment]
Should I add a client to my todo list then?

So far, if I recall correctly, we have:
- Finger daemon running _somewhere_ that pulls information from userdb
- Finger client that contacts finger daemon by default
- inotify daemon that automatically updates information in userdb

**Comment by daurnimator** [comment]
> inotify daemon that automatically updates information in userdb

I wouldn't bother with that.

**Comment by KellerFuchs** [comment]
> Finger daemon running somewhere that pulls information from userdb

In a container running on the same host as `www.hashbang.sh`, to share the IP.

**Comment by KellerFuchs** [comment]
> > inotify daemon that automatically updates information in userdb
> 
> I wouldn't bother with that.

PS: To suggest an alternative UX: `hashbangctl edit finger.foo` dumps `finger.foo` to a temp file, calls `${EDITOR}` on it (default to `/etc/alternatives/editor`), validates and puts the change in userdb once the editor returns.
The nice thing is that it works for whichever thing people build on top of userdb.

**Comment by RyanSquared** [comment]
> > > inotify daemon that automatically updates information in userdb
> > 
> > I wouldn't bother with that.
> 
> PS: To suggest an alternative UX: hashbangctl edit finger.foo dumps finger.foo to a temp file, calls ${EDITOR} on it (default to /etc/alternatives/editor), validates and puts the change in userdb once the editor returns.
> The nice thing is that it works for whichever thing people build on top of userdb.

@lrvick

**Comment by KellerFuchs** [comment]
@ChickenNuggers Are you just pinging @lrvick or did Github eat the text?

**Comment by RyanSquared** [comment]
Was just pinging.
